data/lua/wml/harm_unit.lua

```diff
@@ -160,19 +160,52 @@ function wml_actions.harm_unit(cfg)
 
 			wesnoth.interface.float_label( unit_to_harm.x, unit_to_harm.y, string.format( "<span foreground='red'>%s</span>", text ) )
 
+			local xp_mode = {kill=false, attack=false, defend=false}
+			local experience_split = tostring(experience):split()
+			for _,opt in ipairs(experience_split) do
+				if opt == 'true' or opt == 'yes' or opt == 'nil' then
+					xp_mode.kill = true
+					xp_mode.attack = true
+					xp_mode.defend = true
+				elseif opt == 'fight' then
+					xp_mode.attack = true
+					xp_mode.defend = true
+				elseif opt == 'attack' then
+					xp_mode.attack = true
+				elseif opt == 'defend' then
+					xp_mode.defend = true
+				elseif opt == 'kill' then
+					xp_mode.kill = true
+				elseif not (opt == 'no' or opt == 'false') then
+					wml.error('Invalid [harm_unit] experience, list of values can contain: yes, kill, fight, attack, defend, no')
```

Comment:
I think the error message should look something like this:

```suggestion
					wml.error('Invalid [harm_unit] experience: should be boolean or a list of one or more of the following: kill, fight, attack, defend')
```

It doesn't make sense to mix yes/no as part of a list. We can accept it, but I don't think the error message should be phrased in a way that might encourage it.

Reply:
Lua implementation does not restrict using booleans or duplicates in list. Tests include experience=no,attack,defend,kill. That makes me not want to change schema to more restrictive.

In documentation I planned yes and no still as part of list. Therefore would also keep Lua error according to wiki.

If all addons would be schema-validated at some point during addon lifecycle there would be more motivation to catch invalid configuration before [harm_unit] is used during game.

Wiki: (bold changes since the version since previous PR)
experience: (default yes) If there is a second unit, experience will be attributed similar to regular combat. Can take list of values. When any value is provided, all unspecified experience types default to not giving experience. Supported values:

no - none of involved units get any experience **(this value only makes sense if it is the only value in list)**
kill - the harming unit receives experience based on level of harmed unit in attack which kills harmed unit
attack - the harming unit receives experience based on level of harmed unit in attack which does not kill harmed unit, **or in attack which kills harmed unit while kill experience is disabled**
defend - the unit being harmed receives experience based on level of harming unit in attack which does not kill harmed unit
fight - short form of experience=attack,defend
yes - short form of experience=kill,attack,defend

Reply:
According to your proposed pattern, `yes` is also only valid if it is the only value.

Reply:
I decide to go for {DEFAULT_KEY experience string_list yes} instead of special type.

Reply:
That's also a reasonable approximation, I guess.

Reply:
Balance between cost and benefit. Almost noone knows how to see schema validation results of their addon, so not worth investing too much for schema.

Reply:
> We can accept it, but I don't think the error message should be phrased in a way that might encourage it.

@CelticMinstrel 's suggested change looks like a good change to me too, but a decision either way is what's missing before merging.

Reply:
And who will provide the decision?

Reply:
Please take the change then. :grin: